apis/vmware/v1beta1/vspheremachine_types.go

```diff
@@ -87,6 +87,23 @@ type VSphereMachineSpec struct {
 	//
 	// +optional
 	MinHardwareVersion string `json:"minHardwareVersion,omitempty"`
+
+	// NamingStrategy allows configuring the naming strategy used when calculating the name of the VirtualMachine.
+	// +optional
+	NamingStrategy *VirtualMachineNamingStrategy `json:"namingStrategy,omitempty"`
+}
+
+// VirtualMachineNamingStrategy defines the naming strategy for the VirtualMachines.
+type VirtualMachineNamingStrategy struct {
+	// Template defines the template to use for generating the name of the VirtualMachine object.
+	// If not defined, it will fall back to `{{ .machine.name }}`.
+	// The templating has the following data available:
+	// * `.machine.name`: The name of the Machine object.
```

Comment:
q: should we add something else? thinking about cluster.name, cluster.namespace, virtualMachine.name, vsphereMachine.name

Reply:
Not sure if folks should just freely generate entirely new and different VirtualMachine names. This feature currently assumes that we want to calculate based on the Machine name but we have to make a few changes to it to comply with restrictions.

Reply:
ACK, might be just add a comment that even if template offers some flexibility, we would like the name to somehow link the Machine name for ensuring better user experience when troubleshooting